src/coreclr/nativeaot/System.Private.CoreLib/src/System/Array.NativeAot.cs

```diff
@@ -157,470 +149,109 @@ public unsafe void Initialize()
-            Debug.Assert(destinationArray.ElementMethodTable->IsValueType);
-
-            if (reliable)
-                throw new ArrayTypeMismatchException(SR.ArrayTypeMismatch_CantAssignType);
-
-            MethodTable* destinationElementEEType = destinationArray.ElementMethodTable;
-            nuint destinationElementSize = destinationArray.ElementSize;
-            bool isNullable = destinationElementEEType->IsNullable;
-
-            fixed (byte* pDestinationArray = &MemoryMarshal.GetArrayDataReference(destinationArray))
-            {
-                ref object refSourceArray = ref Unsafe.As<byte, object>(ref MemoryMarshal.GetArrayDataReference(sourceArray));
-                byte* pElement = pDestinationArray + (nuint)destinationIndex * destinationElementSize;
-
-                for (int i = 0; i < length; i++)
-                {
-                    object boxedValue = Unsafe.Add(ref refSourceArray, sourceIndex + i);
-                    if (boxedValue == null)
-                    {
-                        if (!isNullable)
-                            throw new InvalidCastException(SR.InvalidCast_DownCastArrayElement);
-                    }
-                    else
-                    {
-                        MethodTable* eeType = boxedValue.GetMethodTable();
-                        if (!(RuntimeImports.AreTypesAssignable(eeType, destinationElementEEType)))
-                            throw new InvalidCastException(SR.InvalidCast_DownCastArrayElement);
-                    }
-
-                    RuntimeImports.RhUnbox(boxedValue, ref *pElement, destinationElementEEType);
-                    pElement += destinationElementSize;
-                }
+                // Compatible pointers
+                if (RuntimeImports.AreTypesAssignable(sourceElementEEType, destinationElementEEType))
+                    return ArrayAssignType.SimpleCopy;
+                else
+                    return ArrayAssignType.WrongType;
             }
-        }
-
-
-        //
-        // Array.CopyImpl case: Value-type array with embedded gc-references.
-        //
-        private static unsafe void CopyImplValueTypeArrayWithInnerGcRefs(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length, bool reliable)
-        {
-            Debug.Assert(sourceArray.GetMethodTable() == destinationArray.GetMethodTable());
-            Debug.Assert(sourceArray.ElementMethodTable->IsValueType);
-
-            MethodTable* sourceElementEEType = sourceArray.GetMethodTable()->RelatedParameterType;
-            bool reverseCopy = ((object)sourceArray == (object)destinationArray) && (sourceIndex < destinationIndex);
 
-            // Copy scenario: ValueType-array to value-type array with embedded gc-refs.
-            object[]? boxedElements = null;
-            if (reliable)
+            // Different value types
+            if (sourceElementEEType->IsValueType && destinationElementEEType->IsValueType)
```

Comment:
Is this going to prevent the two primitive case from being reached below?

It may be better to structure this like:
```
if (sourceElementEEType->IsValueType || destinationElementEEType->IsValueType)
{
    // Value class boxing
    if (!destinationElementEEType->IsValueType)
    {
...
    }

    // Value class unboxing.
   if (!sourceElementEEType->IsValueType)
   ...

    // Copying primitives from one type to another
    if (sourceElementEEType->IsPrimitive && destinationElementEEType->IsPrimitive)
...

    return ArrayAssignType.WrongType;
}
```